craft_artifacts/artifacts/artifacts.py

```diff
@@ -0,0 +1,132 @@
+# This file is part of craft-artifacts.
+#
+# Copyright 2024 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Artifacts support."""
+
+import abc
+import shutil
+from dataclasses import dataclass, field
+from pathlib import Path
+from typing import cast, final
+
+from pydantic import BaseModel, PrivateAttr
+
+from craft_artifacts.constraints import UniqueList
+
+
+@dataclass
+class BaseArtifactInputDirs:
+    """Base artifact input directories.
+
+    Artifacts classes should expand this to declare their own required input dirs.
+    """
+
+    default_prime_dir: Path
+
+
+@dataclass
+class BaseArtifact(metaclass=abc.ABCMeta):
+    """Base artifact definition."""
+
+    input_dirs: BaseArtifactInputDirs
+    name: str
+    outputs: list[Path] = field(init=False, default_factory=list)
+    work_dir: Path = Path.cwd()
+    keep: bool = False
+
+    def __hash__(self) -> int:
+        return hash(self.name)
+
+    def __eq__(self, other: object) -> bool:
+        if type(other) is type(self):
+            return self.name == cast(BaseArtifact, other).name
```

Comment:
Same question here - shouldn't the inputs and outputs change equality?

Reply:
See my answer above. My goal was to uniquely identify artifacts by their name and make sure we cannot have 2 artifacts with the same name in our collection. In my mind if the outputs of an artifact changes this is still the same artifact.